src/machine-c-api.h

```diff
@@ -82,1894 +83,544 @@ typedef enum { // NOLINT(modernize-use-using)
 /// \param data Receives chunk of memory.
 /// \param length Size of chunk.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_read_virtual_memory(const cm_machine *m, uint64_t address, unsigned char *data, uint64_t length,
-    char **err_msg);
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_read_virtual_memory(const cm_machine *m, uint64_t address, unsigned char *data, uint64_t length);
 
 /// \brief Writes a chunk of data to the machine virtual memory.
-/// \param m Pointer to valid machine instance
+/// \param m Pointer to a valid machine instance.
 /// \param address Virtual address to start writing.
 /// \param data Source for chunk of data.
 /// \param length Size of chunk.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_write_virtual_memory(cm_machine *m, uint64_t address, const unsigned char *data, size_t length,
-    char **err_msg);
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_write_virtual_memory(cm_machine *m, uint64_t address, const unsigned char *data, size_t length);
 
-/// \brief Translates a virtual memory address to its corresponding physical memory address.
-/// \param m Pointer to valid machine instance
-/// \param vaddr Virtual address to translate.
-/// \param paddr Receives the physical memory address.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_translate_virtual_address(cm_machine *m, uint64_t vaddr, uint64_t *paddr, char **err_msg);
+/// \brief Read the value of a word in the machine state.
+/// \param m Pointer to a valid machine instance.
+/// \param address Physical address (aligned to 64-bit boundary).
+/// \param value Receives the value of the word.
+/// \returns 0 for success, non zero code for error.
+/// \warning The current implementation of this function is very slow!
+CM_API int cm_read_word(const cm_machine *m, uint64_t address, uint64_t *value);
+
+/// \brief Write the value of a word in the machine state.
+/// \param m Pointer to a valid machine instance.
+/// \param address Physical address (aligned to 64-bit boundary).
```

Comment:
This function was not there for an obvious reason: the read function can read any word in the state, include those mapped to CSRs and devices. Will this function be able to do the same? See the problem?

Reply:
I imagined, can be tricky to implement. Will remove it then.